api/server/errorhandler.go

```diff
@@ -17,18 +22,99 @@ func makeErrorHandler(err error) http.HandlerFunc {
 		statusCode := httpstatus.FromError(err)
 		vars := mux.Vars(r)
-		if apiVersionSupportsJSONErrors(vars["version"]) {
-			response := &types.ErrorResponse{
-				Message: err.Error(),
-			}
-			_ = httputils.WriteJSON(w, statusCode, response)
-		} else {
+		version := vars["version"]
+
+		const firstAPIVersionWithJSONErrors = "1.24"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONErrors) {
 			http.Error(w, status.Convert(err).Message(), statusCode)
+			return
+		}
+
+		response := marshalErrorResponse(err)
+		const firstAPIVersionWithJSONMultiErrors = "1.44"
+		if version != "" && versions.LessThan(version, firstAPIVersionWithJSONMultiErrors) {
+			response.Errors = nil
+		}
+
+		_ = httputils.WriteJSON(w, statusCode, response)
+	}
+}
+
+type joiningError interface {
+	error
+	Unwrap() []error
+}
+
+type wrappingError interface {
+	error
+	Unwrap() error
+}
+
+// marshalErrorResponse returns a tree of [*types.ErrorResponse] representing the argument err.
+func marshalErrorResponse(err error) *types.ErrorResponse {
+	if errdefs.IsHTTPOnlyError(err) {
+		if wrappedErr := errors.Unwrap(err); wrappedErr != nil {
+			err = wrappedErr
+		}
+	}
+
+	switch err := err.(type) {
+	case joiningError:
+		return marshalJoinedErrors(err)
+	case *multierror.Error:
+		return marshalMultiErrors(err)
+	case wrappingError:
+		return &types.ErrorResponse{
+			Message: err.Error(),
+			Errors:  []*types.ErrorResponse{marshalErrorResponse(err.Unwrap())},
+		}
+	default:
+		return &types.ErrorResponse{
+			Message: err.Error(),
 		}
 	}
 }
 
-func apiVersionSupportsJSONErrors(version string) bool {
-	const firstAPIVersionWithJSONErrors = "1.23"
-	return version == "" || versions.GreaterThan(version, firstAPIVersionWithJSONErrors)
+func formatErrorResponses(errs []*types.ErrorResponse) string {
+	var b strings.Builder
+	b.WriteString(fmt.Sprintf("%d errors occurred:", len(errs)))
+	for _, err := range errs {
+		b.WriteString("\n\t* " + strings.Replace(err.Message, "\n", "\n\t", -1))
+	}
+	return b.String()
+}
+
+func marshalJoinedErrors(err joiningError) *types.ErrorResponse {
+	var subErrors []*types.ErrorResponse
+	for _, subErr := range err.Unwrap() {
+		if subErr != nil {
+			subErrors = append(subErrors, marshalErrorResponse(subErr))
+		}
+	}
+
+	errMsg := err.Error()
+	if len(subErrors) > 1 && err.Error() == errors.Join(err.Unwrap()...).Error() {
+		errMsg = formatErrorResponses(subErrors)
+	} else if len(subErrors) == 1 && errMsg == subErrors[0].Message {
+		return subErrors[0]
+	}
+
+	return &types.ErrorResponse{
+		Message: errMsg,
+		Errors:  subErrors,
+	}
+}
+
+func marshalMultiErrors(err *multierror.Error) *types.ErrorResponse {
+	var subErrors []*types.ErrorResponse
+	for _, subErr := range err.WrappedErrors() {
+		if subErr != nil {
+			subErrors = append(subErrors, marshalErrorResponse(subErr))
+		}
+	}
+
+	return &types.ErrorResponse{
+		Message: formatErrorResponses(subErrors),
```

Comment:
The format of a multierror is customizable. We shouldn't override the intentions of anyone who elected to customize the message format. On the other hand, I have no  objection to overriding _the default way_ multierrors are formatted. [It doesn't render recursive multierrors very nicely.](https://github.com/hashicorp/go-multierror/blob/v1.1.1/format.go#L14)

Reply:
Yeah, that was my initial intent but it turns out setting a custom default format doesn't produce nice errors as it doesn't rely on the recursivity (ie. it doesn't re-use the formatting of children errors). For instance, it'd produce:

```
3 errors occurred:
	* foo
	* 2 errors occurred:
	* bar
	baz
	* one more error
```

Instead of:

```
3 errors occurred:
	* foo
	* 2 errors occurred:
		* bar
		* baz
	* one more error
```

One way to solve that, is to not concatenate `err.Error()`:

```golang
func marshalMultiErrors(err *multierror.Error) *types.ErrorResponse {
	if err.ErrorFormat == nil {
		err.ErrorFormat = func(_ []error) string {
			return formatErrors(subErrors)
		}
	}
}
```

But then, if a custom format was defined, it'll break the formatting defined here or it'll have to reimplement the same logic. So, I'm wondering if we should not unconditionally override the custom format instead.

Reply:
I think we're talking past each other a bit. I'm trying to say that, if the creator of the error assigned a custom format function, we should respect their wishes by using it to render the `ErrorResponse.Message` string. Otherwise I'm fine with using `formatErrors` to do so as the default go-multierror format function `multierror.ListFormatFunc` does not nest nicely. (Explicitly assigning `err.ErrorFormat = multierror.ListFormatFunc` does not count as assigning a custom format function, in my view.)

```go
if err.ErrorFormat != nil && err.ErrorFormat != multierror.ListFormatFunc {
	return &types.ErrorResponse{
		Message: err.Error(),
		Errors: subErrors,
	}
}
return &types.ErrorResponse{
	Message: formatErrors(subErrors),
	Errors: subErrors,
}
```

> But then, if a custom format was defined, it'll break the formatting defined here or it'll have to reimplement the same logic.

```go
func counterpointFormatFunc1(errs []error) string {
	var b strings.Builder
	for i, e := range errs {
		if i > 0 {
			b.WriteString(", ")
		}
		b.WriteString(e.Error())
	}
	return b.String()
}

func counterpointFormatFunc2(errs []error) string {
	return fmt.Sprintf("you've got %d problems", len(errs))
}

func counterpointFormatFunc3(errs []error) string {
	return "there's no rule saying the format func has to interpolate the underlying errors"
}
```

Reply:
OK, I see your point. One nit: it's not possible to compare functions in Go without relying on reflection. So I'll consider that if `err.ErrorFormat` is not nil, it was purposefully defined and we shouldn't dismiss it.